include/cpp-statsd-client/UDPSender.hpp

```diff
@@ -161,27 +163,33 @@ inline UDPSender::~UDPSender() {
         return;
     }
 
-    if (m_batching) {
+    // If we're running a background thread tell it to stop
+    if (m_batchingThread.joinable()) {
         m_mustExit.store(true, std::memory_order_acq_rel);
         m_batchingThread.join();
     }
 
+    // Cleanup the socket
     close(m_socket);
 }
 
 inline void UDPSender::send(const std::string& message) noexcept {
     m_errorMessage.clear();
+
     // If batching is on, accumulate messages in the queue
-    if (m_batching) {
+    if (m_batchsize > 0) {
         queueMessage(message);
         return;
     }
 
+    // Or send it right now
     sendToDaemon(message);
 }
 
 inline void UDPSender::queueMessage(const std::string& message) noexcept {
-    std::unique_lock<std::mutex> batchingLock(m_batchingMutex);
+    // We aquire a lock but only if we actually need to (ie there is a thread also accessing the queue)
+    auto batchingLock =
+        m_batchingThread.joinable() ? std::unique_lock<std::mutex>(m_batchingMutex) : std::unique_lock<std::mutex>();
```

Comment:
we only actually lock if we need to synchronize with the thread, otherwise we get an empty lock which does nothing. this is to make sure that the manual flushing mode is lock-free